### PR TITLE
Examples of radially asymmetric IRFs

### PR DESCRIPTION
**Pull request by AtreyeeS**
After the discussions in #4952 I tried to test some asymmetric IRFs implementation

The good news is that we can support asymmetric aeff and edisp with almost minimal effort.
Defining new IRFs is easy, and we can support building DL4 data products with this PR.

The issue of the PSF is more tricky. We need the `PSFKernelMap`.
Also, the present `PSF` classes are hardcoded to `offset` axis, so that will need some adaptation.

If we decide to support these classes internally, we would need a good inheritance scheme to avoid code duplication. All `IRF3D` classes will have some common functionalities like `plot_at_energy` and `to_2d`. Similarly for all `IRF2D` classes.
Also, having `EffectiveAreaIRF` `EnergyDispersionIRF` classes like our present `BackgroundIRF` will be good.

We also need a defined format to understand where to raise warnings about the IRFs. In this PR, I kept the old behaviour and raise them only for the Background classes, because there is indeed no point in raising the same issue for all the maps.

comments @maxnoe @bkhelifi @registerrier @QRemy @adonath @LauraOlivera

### CODE REVIEW

File: examples/tutorials/api/irfs.py [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+# -----------------
+#
+
+
+class EnergyDispersion3D(IRF):
+    tag = "edisp_3d"
+    required_axes = ["energy_true", "migra", "fov_lon", "fov_lat"]
+    default_unit = u.one
+
+    @classmethod
```
Comment (registerrier) [inline examples/tutorials/api/irfs.py:114]: you might not expose this in a tutorial as this is quite complex. Showing a simple diagonal one would be enough.

File: examples/tutorials/api/irfs.py [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+# reduction
+#
+
+import numpy as np
+import scipy.special
+import astropy.units as u
+from astropy.coordinates import SkyCoord
+from gammapy.irf import *
+from gammapy.irf.io import COMMON_IRF_HEADERS, IRF_DL3_HDU_SPECIFICATION
+from gammapy.makers.utils import *
```
Comment (registerrier) [inline examples/tutorials/api/irfs.py:25]: not needed here

File: gammapy/irf/background.py [hunk 3]
```diff
@@ -11,7 +11,7 @@
 from .core import IRF
 from .io import gadf_is_pointlike
 
-__all__ = ["Background3D", "Background2D"]
+__all__ = ["Background3D", "Background2D", "BackgroundIRF"]
```
Comment (registerrier) [inline gammapy/irf/background.py:14]: Why do you want to expose `BackgroundIRF`? Because of documentation?

File: gammapy/makers/utils.py [hunk 4]
```diff
@@ -28,6 +28,74 @@
 log = logging.getLogger(__name__)
 
 
+def _get_fov_coords(pointing, irf, geom, use_region_center=True, obstime=None):
```
Comment (registerrier) [inline gammapy/makers/utils.py:31]: Is it the right place for this function? It could also live in pointing or irf.

**Review by registerrier — COMMENTED** [review]
Thanks @AtreyeeS . This is nice!

Just a few rapid comments for a start:
- I think this should be two distinct PRs:
  - the first one should introduce and test support for the different FoV systems (offset or lon,lat) in the `Makers`
  - the second would then introduce the documentation for the IRF
- Regarding the former, would it be worth creating functions such `coord_to_fov_lonlat` and `coord_to_fov_offset` that could take a `FoValignement` as argument in order to have it independent from `IRF`?
- For the tutorial, I think the content could be lighter, e.g. I would not expose a `from_gauss`. Maybe it misses also a general discussion on how IRFs are implemented with some examples of regular IRFs.

File: examples/tutorials/api/irfs.py (hunk 1 above)
Comment (AtreyeeS) [reply examples/tutorials/api/irfs.py:114]: Indeed! This is just to show that most of the functions can be easily generalised to the 3D case.

File: gammapy/irf/background.py (hunk 3 above)
Comment (maxnoe) [reply gammapy/irf/background.py:14]: I think presence in `__all__` is required for it to show up in the API documentation.

I think the best way here would be to expand the docstring of the IRF baseclasses to explain to developers what is necessary to implement a new implementation.

**Comment by AtreyeeS** [comment]
Thanks @registerrier ! My main aim here was to show that 

>   * the first one should introduce and test support for the different FoV systems (offset or lon,lat) in the `Makers`
>   * the second would then introduce the documentation for the IRF

If we want to test 3D irfs, it would mean adding these classes in the test files and then testing.
To test the makers, it would be better to have some 3D IRFs to use for testing (@maxnoe do you have some examples?). Otherwise, we can create some from the standard IRFs by rotating around the offset axis.

Or shall we directly add support for 3D IRFs in our main code base? That would need a decision on the inheritance scheme, for ex. And a defined format.

**Comment by maxnoe** [comment]
> To test the makers, it would be better to have some 3D IRFs to use for testing (@maxnoe do you have some examples?). 

Not yet, but I have a new master's student who is now starting implementation of 3d effective area and energy dispersion in pyirf.

**Comment by bkhelifi** [comment]
As the `BackgroundIRF' is now exposed into the documentation, it might be good to improve the docstring, by explaining the content of this base class

File: gammapy/makers/utils.py (hunk 4 above)
Comment (bkhelifi) [reply gammapy/makers/utils.py:31]: @AtreyeeS: should we then follow these recommendations ?

File: examples/tutorials/api/irfs.py [hunk 5]
```diff
@@ -0,0 +1,458 @@
...
+        migra_max = coords["migra"][:, 1:, :]
+
+        # Analytical formula for integral of Gaussian
+        s = np.sqrt(2) * sigma
+        t1 = (migra_max - 1 - bias) / s
+        t2 = (migra_min - 1 - bias) / s
+        pdf = (scipy.special.erf(t1) - scipy.special.erf(t2)) / 2
+        pdf = pdf / (migra_max - migra_min)
+
+        r1 = np.rollaxis(pdf, -1, 1)
```
Comment (bkhelifi) [inline examples/tutorials/api/irfs.py:262]: As this function is exposed, can you add a simple comment to explain what do these 2 lines?

File: examples/tutorials/api/irfs.py [hunk 6]
```diff
@@ -0,0 +1,458 @@
...
+migra = np.array([0.98, 0.97, 0.7])
+fov_lon = [0.1, 1.5] * u.deg
+fov_lat = [0.0, 0.3] * u.deg
+
+edisp3d.evaluate(
+    energy_true=energy.reshape(-1, 1, 1, 1),
+    migra=migra.reshape(1, -1, 1, 1),
+    fov_lon=fov_lon.reshape(1, 1, -1, 1),
+    fov_lat=fov_lat.reshape(1, 1, 1, -1),
+)[0][2]
```
Comment (bkhelifi) [inline examples/tutorials/api/irfs.py:342]: Can you explain the [0][2] to the user?
Can you precise also how the interpolations over the axis and the data are made? Or make a link towards an other tutorial...

File: examples/tutorials/api/irfs.py [hunk 7]
```diff
@@ -0,0 +1,458 @@
...
+# While Gammapy does not have inbuilt classes for supporting asymmetric
+# IRFs (except for `Background3D`), custom classes can be created. For
+# this to work correctly with the `MapDatasetMaker`, only variations
+# with `fov_lon` and `fov_lat` can be allowed.
+#
+# The main idea is that the list of required axes should be correctly
+# mentioned in the class definition.
+#
+# Asymmetric PSFs are not correctly supported at present in the data
+# reduction scheme.
```
Comment (bkhelifi) [inline examples/tutorials/api/irfs.py:126]: Maybe you can add that one can however create an asymmetric PSF with the current API, as exemplified here... In this way, the user will not be surprised to see that in the tuto

File: examples/tutorials/api/irfs.py [hunk 8]
```diff
@@ -0,0 +1,458 @@
...
+######################################################################
+# PSF
+# ---
+#
+# A higher dimensional PSF Table can also be created, as shown here.
+# Support for higher dimensional analystic PSF will come later
+#
+
+
+class PSFnD(IRF):
```
Comment (bkhelifi) [inline examples/tutorials/api/irfs.py:393]:
```suggestion
class PSF4D(IRF):
```

**Review by bkhelifi — CHANGES_REQUESTED** [review]
Thanks a lot @AtreyeeS ... I have made some suggestions

File: examples/tutorials/api/irfs.py [hunk 9]
```diff
@@ -0,0 +1,458 @@
...
+# The peek methods give quick looks into the irfs
+aeff.peek()
+
+
+######################################################################
+# Similarly, we can access other IRFs as well
+#
+
+psf = EnergyDependentMultiGaussPSF.read(irf_filename, hdu="Point Spread Function")
+bkg = Background3D.read(irf_filename, hdu="BACKGROUND")
```
[the remark this answers is not in the document]
Comment (AtreyeeS) [reply examples/tutorials/api/irfs.py:68]: The CTA IRFs are in 3D. They are actually symmetric when you plot them, but are supplied as per `Background3D` specifications. If I remember correctly a conversation with Gernot, the IRFs were initially generated in `energy_true` and `offset`, but then supplied as 3D

File: examples/tutorials/api/irfs.py (hunk 5 above)
Comment (AtreyeeS) [reply examples/tutorials/api/irfs.py:262]: Removed now

**Comment by AtreyeeS** [comment]
Thanks @bkhelifi for the detailed review. I addressed most of your comments, can you take a look again please?

File: examples/tutorials/api/irfs.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (maxnoe) [reply examples/tutorials/api/irfs.py:68]: Public CTAO IRFs should be 2d for Prod3, and pseudo-3d for Prob5.

File: examples/tutorials/api/irfs.py [hunk 10]
```diff
@@ -0,0 +1,495 @@
...
+#
+
+bkg.interp_kwargs
+
+# Evaluate background
+energy = [1, 10, 100, 1000] * u.TeV
+fov_lon = [
+    1,
+    2,
+] * u.deg
```
Comment (registerrier) [inline examples/tutorials/api/irfs.py:98]: Maybe try to format it as the line below

File: examples/tutorials/api/irfs.py [hunk 11]
```diff
@@ -0,0 +1,495 @@
...
+fov_lat = [1, 2] * u.deg
+ev = bkg.evaluate(
```
Comment (registerrier) [inline examples/tutorials/api/irfs.py:100]: Maybe add a comment explaining that evaluation supports broadcasting numpy-style

As the command might look cryptic

File: examples/tutorials/api/irfs.py [hunk 12]
```diff
@@ -0,0 +1,495 @@
...
+# ~~~~~~~~~~~~~
+#
+
+IRF_DL3_HDU_SPECIFICATION["psf_nd"] = {
+    "extname": "POINT SPREAD FUNCTION",
+    "column_name": "MATRIX",
+    "mandatory_keywords": {
+        **COMMON_IRF_HEADERS,
+        "HDUCLAS2": "PSF",
+        "HDUCLAS3": "FULL-ENCLOSURE",  # added here to have HDUCLASN in order
```
Comment (registerrier) [inline examples/tutorials/api/irfs.py:478]:
```suggestion
        "HDUCLAS3": "FULL-ENCLOSURE",
```

File: examples/tutorials/api/irfs.py [hunk 13]
```diff
@@ -0,0 +1,495 @@
...
+# ---
+#
+# Asymmetric PSFs are not correctly supported at present in the data
+# reduction scheme. However, higher dimensional PSF Table can be created
+# as a data container, as shown here.
+#
+
+
+class PSF_assym(IRF):
+    tag = "psf_nd"
```
Comment (registerrier) [inline examples/tutorials/api/irfs.py:428]: Maybe use `psf_asym` for clarity?

File: examples/tutorials/api/irfs.py [hunk 14]
```diff
@@ -0,0 +1,495 @@
...
+    fov_lat=fov_lat.reshape(1, 1, 1, -1),
+)
+
+
+######################################################################
+# Serialisation
+# ~~~~~~~~~~~~~
+#
+
+IRF_DL3_HDU_SPECIFICATION["psf_nd"] = {
```
Comment (registerrier) [inline examples/tutorials/api/irfs.py:472]: `"psf_asym"`?

File: gammapy/makers/utils.py (hunk 4 above)
Comment (registerrier) [reply gammapy/makers/utils.py:31]: For now, I would suggest adding a TODO and open an issue about this.

**Review by registerrier — APPROVED** [review]
Thanks @AtreyeeS . I have left a few minor comments, in particular on the tutorial.

File: gammapy/makers/utils.py (hunk 4 above)
Comment (AtreyeeS) [reply gammapy/makers/utils.py:31]: Created #5041

**Review by registerrier — APPROVED** [review]
Thanks @AtreyeeS . This is ready to merge. No further comment.

**Comment by registerrier** [comment]
CI fail is unrelated. Merging.

**Pull request merged**